compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -46,30 +46,81 @@ type Res = def::Res<NodeId>;
 impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
     /// Defines `name` in namespace `ns` of module `parent` to be `def` if it is not yet defined;
     /// otherwise, reports an error.
-    pub(crate) fn define_binding(
+    pub(crate) fn define_binding_local(
         &mut self,
         parent: Module<'ra>,
         ident: Ident,
         ns: Namespace,
         binding: NameBinding<'ra>,
     ) {
-        if let Err(old_binding) = self.try_define(parent, ident, ns, binding, false) {
+        assert!(parent.is_local());
+        if let Err(old_binding) = self.try_define_local(parent, ident, ns, binding, false) {
             self.report_conflict(parent, ident, ns, old_binding, binding);
         }
     }
 
-    fn define(
+    fn define_local(
         &mut self,
         parent: Module<'ra>,
         ident: Ident,
         ns: Namespace,
         res: Res,
-        vis: Visibility<impl Into<DefId>>,
+        vis: Visibility, // Implicitly local
         span: Span,
         expn_id: LocalExpnId,
     ) {
+        assert!(parent.is_local());
+        assert!(res.opt_def_id().is_none_or(|def_id| def_id.is_local()));
         let binding = self.arenas.new_res_binding(res, vis.to_def_id(), span, expn_id);
-        self.define_binding(parent, ident, ns, binding)
+        self.define_binding_local(parent, ident, ns, binding);
+    }
+
+    /// Defines `name` in namespace `ns` of module `parent` to be `def` if it is not yet defined;
+    /// otherwise, panic.
+    pub(crate) fn define_binding_extern(
+        &self,
+        parent: Module<'ra>,
+        ident: Ident,
+        ns: Namespace,
+        binding: NameBinding<'ra>,
+    ) {
+        assert!(!parent.is_local());
+        // Even if underscore names cannot be looked up, we still need to add them to modules,
+        // because they can be fetched by glob imports from those modules, and bring traits
+        // into scope both directly and through glob imports.
+        let key = BindingKey::new_disambiguated(ident, ns, || {
+            parent.underscore_disambiguator.update(|d| d + 1);
+            parent.underscore_disambiguator.get()
+        });
+        let resolution = &mut *self.resolution(parent, key).borrow_mut();
+        let resolution_binding = if binding.is_glob_import() {
+            &mut resolution.glob_binding
+        } else {
+            &mut resolution.non_glob_binding
+        };
+        if resolution_binding.replace(binding).is_some() {
```

Comment:
```suggestion
        if resolution_binding.insert(binding).is_some() {
```
The exact behavior doesn't matter because we panic, but `insert` is a more common method.

Reply:
This doesn't give me the ability to check if a resolution was already present. Since collisions should not happen here.

Reply:
Hm? `insert` also returns whether the old key already existed in the map.
See e.g. `fn record_partial_res`.

Reply:
Signature of `Option::insert` is:
```rust
pub fn insert(&mut self, value: T) -> &mut T { ... }
```

Where `&mut T` is the new value, regardless of whether a value already resides in the Option.

`Option::replace` returns `Option<T>` indicating the replaced value.

Reply:
I think you are confused with `HashMap::insert`, which has the behaviour of `Option::replace`.

Reply:
Ah, right, I thought it was a set/map.